### PR TITLE
Fix GitHub snippet resolution

### LINKED ISSUE #771: Faulty snippet detection on GitHub issues
- Visit https://github.com/gorilla/mux/issues/381
- Notice hovers don't work on the first snippet

It looks like [this code](https://sourcegraph.sgdev.org/github.com/sourcegraph/browser-extensions@71791ff24f6f87ff86b85008ea3405c0e49843ca/-/blob/src/libs/github/file_info.ts#L92) assumes that the first `<a>` tag within a comment that links to a URL that can be parsed by `getGitHubState(anchor.href)` is a snippet, but in that issue it's actually a link to a commit.

image

image

One solution is to not stop after the first is found, and to tighten the URL regex.

### LINKED ISSUE #3105: GitHub snippet codeintel is broken
Go to https://github.com/sourcegraph/sourcegraph/issues/3085#L133, hover over tokens in the snippet.

Console error:

```
file_info.ts:72 Uncaught Error: Unable to determine the file path of the current file because the current URL (window.location https://github.com/sourcegraph/sourcegraph/issues/3085#L133) does not have a file path.
    at Object.resolveFileInfo (file_info.ts:72)
    at MergeMapSubscriber.project (code_intelligence.tsx:488)
    at MergeMapSubscriber._tryNext (mergeMap.ts:141)
    at MergeMapSubscriber._next (mergeMap.ts:132)
    at MergeMapSubscriber.Subscriber.next (Subscriber.ts:100)
    at MergeMapSubscriber.notifyNext (mergeMap.ts:170)
    at InnerSubscriber._next (InnerSubscriber.ts:18)
    at InnerSubscriber.Subscriber.next (Subscriber.ts:100)
    at MergeMapSubscriber.notifyNext (mergeMap.ts:170)
    at InnerSubscriber._next (InnerSubscriber.ts:18)
```

Seems like code some part of the code only expects single file blob pages.

### LINKED ISSUE #3110: Comments detected as code views on GitHub
image

### PR DESCRIPTION
**Pull request by lguychard**
Fixes #771
Fixes #3105
Fixes #3110

Deprecates `commentSnippetCodeViewResolver`, which:
- Matched entire comments, even though they may have contained several snippets (#3110)
- Used the first link whose href could be parsed by `getGitHubState()`, which could be any inline link in a GitHub comment (#771)

Snippet code views are now resolved using the `fileLineContainerResolver`, to the closes element containing both the code and the permalink. Several snippets contained in the same comment can be accurately resolved.

##### Code Hosts

- [x] GitHub
- [x] GitHub Enterprise
- [ ] Refined GitHub
- [ ] Phabricator
- [ ] Phabricator integration
- [ ] Bitbucket
- [ ] Gitlab

##### Browsers

- [x] Chrome
- [ ] Firefox

image
image

### CODE REVIEW

**Comment by felixfbecker** [comment]
What do you mean with "deprecates"? Why can't we remove it?

**Comment by lguychard** [comment]
@felixfbecker poor wording on my part, it is removed, the useful bits are kept in `snippetCodeView`.

File: browser/src/libs/github/code_intelligence.ts [hunk 1]
```diff
@@ -154,12 +153,24 @@ export const createFileLineContainerToolbarMount: NonNullable<CodeView['getToolb
...
     resolveView: (fileLineContainer: HTMLElement): CodeView | null => {
+        const embeddedBlobWrapper = fileLineContainer.closest('.blob-wrapper-embedded')
+        if (embeddedBlobWrapper) {
+            // This is a snippet embedded in a comment.
+            // Resolve to `.blob-wrapper-embedded`'s parent element,
+            // the smallest element that contains both the code and
+            // the HTML anchor allowing to resolve the file info.
+            const element = embeddedBlobWrapper.parentElement!
+            return {
+                element,
```
Comment (felixfbecker) [inline browser/src/libs/github/code_intelligence.ts]: I think we can't return a different element than what was found with the selector. We don't run the resolver for removed code views. If an element found by the selector was removed, we look it up in the Map to find its code view state. Previously, that code view state was stored with `CodeView.element` as the key. So if they differ, the code view state won't get unsubscribed from and we will hold on to it, its editors etc forever

Comment (lguychard) [reply browser/src/libs/github/code_intelligence.ts]: You're right, but this means the `fileLineContainerResolver` was already causing leaked views prior to this PR by returning a different element. In both cases (returning the element for the blob code view, or the snippet code view), there's no selector that makes sense that would really match _just_ the code view. Maybe we should allow resolvers like this, and change our handling of removed nodes?

Comment (felixfbecker) [reply browser/src/libs/github/code_intelligence.ts]: How about we change `selector: string` to `getElements: (container: HTMLElement) => Iterable<HTMLElement>`

Comment (lguychard) [reply browser/src/libs/github/code_intelligence.ts]: PTAL https://github.com/sourcegraph/sourcegraph/pull/3976 and let me know what you think.

**Comment by felixfbecker** [comment]
@lguychard I think this is fine to merge since it doesn't regress on anything, it's a net improvement, right?

**Comment by lguychard** [comment]
Yeah, I was holding off for https://github.com/sourcegraph/sourcegraph/pull/3976, but we can merge without it.

**Pull request merged**